File: web_viewer/scripts/compile-authoritative-story-candidate.py
"""Compile one compatibility scenario JSON to Python-native authoritative v2."""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
PIPELINE_ROOT = ROOT.parent / "data_pipeline"
sys.path.insert(0, str(PIPELINE_ROOT))

from scenario_compiler import ScenarioCompiler  # noqa: E402


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", required=True)
    parser.add_argument("--output", required=True)
    parser.add_argument("--compiler-version", default="scenario-compiler-python-v2")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    input_path = Path(args.input).resolve()
    output_path = Path(args.output).resolve()
    if output_path == ROOT or ROOT in output_path.parents:
        raise ValueError("Candidate output must be outside the web_viewer workspace")
    with input_path.open("r", encoding="utf-8") as handle:
        compatibility = json.load(handle)
    authoritative = ScenarioCompiler.to_authoritative(compatibility, args.compiler_version)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with output_path.open("w", encoding="utf-8", newline="\n") as handle:
        json.dump(authoritative, handle, ensure_ascii=False, indent=2)
        handle.write("\n")
    print(f"Python authoritative candidate: {output_path} ({len(authoritative['steps'])} steps)")


if __name__ == "__main__":
    main()

File: web_viewer/scripts/compile-authoritative-story-candidate.mjs
import { readFile, writeFile } from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { compileAuthoritativeScenario } from './lib/authoritative-scenario-compiler.mjs'

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..')
const args = Object.fromEntries(process.argv.slice(2).map(value => {
  const [key, ...rest] = value.replace(/^--/, '').split('=')
  return [key, rest.join('=')]
}))
if (!args.input || !args.output) {
  throw new Error('Usage: node scripts/compile-authoritative-story-candidate.mjs --input=<json> --output=<json> [--compiler-version=<version>]')
}
const inputPath = path.resolve(args.input)
const outputPath = path.resolve(args.output)
const relativeOutput = path.relative(root, outputPath)
if (relativeOutput === '' || (!relativeOutput.startsWith('..') && !path.isAbsolute(relativeOutput))) {
  throw new Error('Authoritative candidate output must remain outside the web_viewer workspace')
}
const input = JSON.parse(await readFile(inputPath, 'utf8'))
const output = compileAuthoritativeScenario(input, { compilerVersion: args['compiler-version'] })
await writeFile(outputPath, `${JSON.stringify(output, null, 2)}\n`, 'utf8')
console.log(JSON.stringify({ input: inputPath, output: outputPath, steps: output.steps.length }, null, 2))
